### PR TITLE
Add alignment snapping for auras and groups

### PR DESCRIPTION
**Pull request by mrbuds**
# Description
Add alignment snapping for auras and groups
Pressing shift ignore the snapping

https://i.imgur.com/Yl0rKDs.gif

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [x] snapping aura
- [x] snapping group

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by mrbuds** [comment]
Fixed for groups, and made group ignore snapping with it's children

Do you think the shift key for ignoring snapping need "something" to show that you can use it ? or is it common enough on editing software to let users find out themselves ? Or should it be the other way around, you need to press a key to activate snapping ?

**Comment by emptyrivers** [comment]
Enable it by default, imo. If we had a preferences panel, then I’d be comfortable dropping a setting for that in there (we don’t need a row of checkboxes across the top of the options panel for every little one-off setting).

**Comment by Stanzilla** [comment]
+1 for default enable. You can leave shit holding to disable it, probably does not need that much exposure and we can mention it in the changelog.

**Comment by mrbuds** [comment]
ok that's current behavior

**Comment by emptyrivers** [comment]
@Stanzilla why is this an epic?

**Comment by Stanzilla** [comment]
> @Stanzilla why is this an epic?

because I did not read what the label text says and thought it's epic!

**Comment by emptyrivers** [comment]
😂

File: WeakAurasOptions/OptionsFrames/MoverSizer.lua [hunk 1]
```diff
@@ -450,6 +464,32 @@ local function ConstructMoverSizer(parent)
...
+        if k ~= data.id and v.view.visibility ~= 0 then
+          tinsert(x, v.view.region:GetLeft())
+          tinsert(x, v.view.region:GetRight())
+          tinsert(y, v.view.region:GetTop())
+          tinsert(y, v.view.region:GetBottom())
+        end
+      end
+      table.sort(x)
+      table.sort(y)
+      for k, v in ipairs(x) do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/MoverSizer.lua]: the convention is to use k,v for "key", "value" that is "pairs".

File: WeakAurasOptions/OptionsFrames/MoverSizer.lua [hunk 2]
```diff
@@ -450,6 +464,32 @@ local function ConstructMoverSizer(parent)
...
+        if v ~= x[k+1] then
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/MoverSizer.lua]: So here, we probably want to do something like, if abs(v - x[k+1]) > 2 then

Basically throwing away anything that is almost the same as the next coordinate.

Comment (InfusOnWoW) [reply WeakAurasOptions/OptionsFrames/MoverSizer.lua]: On thinking again, nah what you have is actually fine.

File: WeakAurasOptions/OptionsFrames/MoverSizer.lua [hunk 3]
```diff
@@ -98,7 +98,21 @@ local function ConstructMover(frame)
   right:GetPushedTexture():SetRotation(-math.pi/2)
   right:SetScript("OnClick", function() moveOnePxl("right") end)
 
-  return top, bottom, left, right
+  local linex = frame:CreateLine(nil, "OVERLAY", 7)
```
Comment (Stanzilla) [inline WeakAurasOptions/OptionsFrames/MoverSizer.lua]: lineX maybe?

**Comment by mrbuds** [comment]
These last minute commits adds:
- Snapping will search for matching frames in reverse order if the mouse is moved in left or bottom direction
It should make positioning feel more natural when there are 2 close matching frames
- Added horizontal and vertical lines for middle of UIParent https://i.imgur.com/K7SOQ5y.jpg
- Holding ctrl will align with middle of frame instead of sides
But there is a little trick with the Ctrl key, the list of snapping lines is build when you start moving frames (mousedown)
If you don't hold Ctrl on mousedown the snapping will center your aura with sides of visible auras https://i.imgur.com/o4BSiFa.png
If you hold Ctrl on mousedown, the snapping will center your aura with middle of visible auras https://i.imgur.com/zgtXjIQ.png

**Comment by mrbuds** [comment]
Added selection of auras by clicking directly on their displayed region, support ctrl for multi selection.

TODO: (for 2.13 ?)
- make resizing use snapping lines
- handle multi-selection for moving & resizing

**Comment by InfusOnWoW** [comment]
I've found two bugs so far:

1) 
* Create a new icon
* Create a second icon
* Make sure both are visible by e.g. clicking on the loaded eye
* Select them both by clicking on the actual auras
=> In the list view both are shown as selected, but changing any settings only affects one aura

2)
* Create a new icon
* Click the eye icon next to it and drag it to a new place
* Create a second new icon
* Drag it away
=> Clicking on the first created aura doesn't select it

**Comment by InfusOnWoW** [comment]
Another small bug:
https://youtu.be/svnewt6nAgE

As I drop the middle icon towards the left or right line, it disappears and the other line appears. And as I drag towards that line, it disappears.

**Comment by InfusOnWoW** [comment]
The OnMouseDown handled remains after closing the options.
Steps to reproduce:
* Add a "print" to the handler for OnMouseDown
* Open the options, Close the Options
* Click on a aura

**Comment by InfusOnWoW** [comment]
We also break some auras if they for some reason set a script, but I'm entirely okay with that.

**Comment by mrbuds** [comment]
> 
> 
> I've found two bugs so far:
> 
> 
>     * Create a new icon
> 
>     * Create a second icon
> 
>     * Make sure both are visible by e.g. clicking on the loaded eye
> 
>     * Select them both by clicking on the actual auras
>       => In the list view both are shown as selected, but changing any settings only affects one aura
> 
> 
> 
>     * Create a new icon
> 
>     * Click the eye icon next to it and drag it to a new place
> 
>     * Create a second new icon
> 
>     * Drag it away
>       => Clicking on the first created aura doesn't select it

fixed

**Comment by InfusOnWoW** [comment]
Pressing Shift while dragging seems to cancel the dragging for me.

**Comment by InfusOnWoW** [comment]
> if event == "MODIFIER_STATE_CHANGED" and key == "LCTRL" or key == "RCTRL" then

In lua, and binds stronger than or so this is:

if (event == "MODIFIER_STATE_CHANGED" and key == "LCTRL") or key == "RCTRL" then

File: WeakAurasOptions/OptionsFrames/OptionsFrame.lua [hunk 4]
```diff
@@ -661,21 +661,21 @@ function WeakAuras.CreateFrame()
     self:FillOptions(displayOptions[tempGroup.id]);
   end
 
-  frame.ClearPicks = function(self, except)
+  frame.ClearPicks = function(self, noHide)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/OptionsFrame.lua]: There's a caller that calls ClearPicks with a aura name: 
WeakAuras.ClearPicks(data.id);

**Comment by mrbuds** [comment]
> 
> 
> The OnMouseDown handled remains after closing the options.
> Steps to reproduce:
> 
>     * Add a "print" to the handler for OnMouseDown
> 
>     * Open the options, Close the Options
> 
>     * Click on a aura

https://github.com/WeakAuras/WeakAuras2/pull/1254/commits/175e96fcfa8f0cded73713cb6e6841ca2a483196 fix it but it feels dirty, do you see a better way to do it?

**Comment by InfusOnWoW** [comment]
I think that's okay.

File: WeakAurasOptions/OptionsFrames/OptionsFrame.lua (hunk 4 above)
Comment (mrbuds) [reply WeakAurasOptions/OptionsFrames/OptionsFrame.lua]: "except" is not used by the function

Comment (mrbuds) [reply WeakAurasOptions/OptionsFrames/OptionsFrame.lua]: or i'm blind

Comment (InfusOnWoW) [reply WeakAurasOptions/OptionsFrames/OptionsFrame.lua]: That doesn't mean you should fix the call that passes a parameter!

**Comment by mrbuds** [comment]
Everything fixed, i also removed the ability to use ctrl for multi-selection directly on regions until mover/sizer handle it

**Comment by Stanzilla** [comment]
it's time for the obligatory "bump the minor" comment

**Pull request merged**